controllers/configentry_controller.go

```diff
@@ -0,0 +1,299 @@
+	AddFinalizer(string)
+	// RemoveFinalizer removes this finalizer from the list.
+	RemoveFinalizer(string)
+	// Finalizers returns the list of finalizers for this object.
+	Finalizers() []string
+	// Kind returns the Consul config entry kind, i.e. service-defaults, not
+	// ServiceDefaults.
+	Kind() string
+	// Name returns the name of the config entry.
+	Name() string
+	// SetConditions updates conditions.
+	SetConditions(consulv1alpha1.Conditions)
+	// GetCondition returns condition of this type.
+	GetCondition(consulv1alpha1.ConditionType) *consulv1alpha1.Condition
+	// ToConsul converts the CRD to the corresponding Consul API definition.
+	// Its return type is the generic ConfigEntry but a specific config entry
+	// type should be constructed e.g. ServiceConfigEntry.
+	ToConsul() capi.ConfigEntry
+	// MatchesConsul returns true if the CRD has the same fields as the Consul
+	// config entry.
+	MatchesConsul(capi.ConfigEntry) bool
+	// GetObjectKind should be implemented by the generated code.
+	GetObjectKind() schema.ObjectKind
+	// DeepCopyObject should be implemented by the generated code.
+	DeepCopyObject() runtime.Object
+	// Validate returns an error if the CRD is invalid.
+	Validate() error
+}
+
+// ConfigEntryReconciler is a generic reconciler that is used to reconcile
+// all config entry types, e.g. ServiceDefaults, ServiceResolver, etc, since
+// they share the same reconcile behaviour.
+type ConfigEntryReconciler struct {
+	ConsulClient *capi.Client
+
+	// EnableConsulNamespaces indicates that a user is running Consul Enterprise
+	// with version 1.7+ which supports namespaces.
+	EnableConsulNamespaces bool
+
+	// ConsulDestinationNamespace is the name of the Consul namespace to create
+	// all config entries in. If EnableNSMirroring is true this is ignored.
+	ConsulDestinationNamespace string
+
+	// EnableNSMirroring causes Consul namespaces to be created to match the
+	// k8s namespace of any config entry custom resource. Config entries will
+	// be created in the matching Consul namespace.
+	EnableNSMirroring bool
+
+	// NSMirroringPrefix is an optional prefix that can be added to the Consul
+	// namespaces created while mirroring. For example, if it is set to "k8s-",
+	// then the k8s `default` namespace will be mirrored in Consul's
+	// `k8s-default` namespace.
+	NSMirroringPrefix string
+
+	// CrossNSACLPolicy is the name of the ACL policy to attach to
+	// any created Consul namespaces to allow cross namespace service discovery.
+	// Only necessary if ACLs are enabled.
+	CrossNSACLPolicy string
+}
+
+// Reconcile reconciles Kubernetes' state with Consul's. CRD-specific reconciler's
+// call this function because it handles reconciliation of config entries
+// generically.
+// CRD-specific reconcilers should pass themselves in as updater since we
+// need to call back into their own update methods to ensure they update their
+// internal state.
+func (r *ConfigEntryReconciler) Reconcile(
```

Comment:
Can we also change the name to `ReconcileEntry` just to ensure this is not confused for being something that implements the `controller-runtime` Reconciler interface?

Reply:
Done.